Issue/13326 scan fixing in builder

ashiagr opened the pull request:
Parent #13326

This PR moves threat fixing state item building logic to the builder class. It fixes the issue when progress bar was not displayed when an ignored threat was fixed.

It also 
- updates header, description
- removes labels from the progress bar 
- adds loading spinner on threat list items when threats are fixing
to have a bit of similarity with the iOS app.

https://user-images.githubusercontent.com/1405144/106699637-4303e680-6609-11eb-8baf-a32cc766b85f.mov

**To test**

Prerequisite:

- A threat is added to the testing site. In case you run out of threats, re-install bad calendar plugin and re-run scan.
- Scan feature flag is enabled

1. Open Scan screen
2. Start scan
3. Click on "Fix threats" button
4. Notice that 
    - app displays "Fixing Threats" state and progress indicator is displayed on the threat items
    - on pressing back and returning to the screen while threats are fixing, "Fixing Threats" state is re-shown
    - as threats get fixed, they are removed from the list
    - once fixing threats list is empty, scan state is re-fetched and displayed

** Also test the flow when an ignored threat is fixed with steps from PR #13924.

**Notes**
- Ignore styling

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ashiagr committed:
Move threat fixing state building into builder class

ashiagr committed:
Tests: Fix existing tests to compile

ashiagr committed:
Add clickable text info in the description state

ashiagr committed:
Pass clickable text info from the builder

ashiagr committed:
Show clickable text in description viewolder

ashiagr committed:
Tests: Add test for clickable text in scan state model with threats

ashiagr committed:
Fix ktlint, remove trailing comma

ashiagr committed:
Replace threats fixing description text

ashiagr committed:
Merge branch 'develop' into issue/13326-scan-fixing-in-builder

ashiagr committed:
Tests: Add tests for fixing threats state

ashiagr committed:
Remove unused progress bar from threats found scan state

ashiagr committed:
Get threat from scan store instead of scan state model for fixing threats state

This helps showing fixing threat state when threat does not belong to the scan state model, e.g. when fixing an ignored threat.

ashiagr committed:
Remove threat found header

ashiagr committed:
Show loading instead of icon if threat is fixing

ashiagr committed:
Refactor: Tests - use expected values directly in assert

ashiagr committed:
Tests: Add tests for loading/ icon visibility

ashiagr committed:
Show fixable threat description in fixing threat item sub header

ashiagr committed:
Merge branch 'develop' into issue/13326-scan-fixing-in-builder

ashiagr committed:
Tests: Fix failing tests

ashiagr committed:
Merge branch 'issue/13326-scan-fixing-in-builder' into issue/13326-here-to-help

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/jetpack/scan/builders/ScanStateListItemsBuilder.kt
#	WordPress/src/test/java/org/wordpress/android/ui/jetpack/scan/builders/ScanStateListItemsBuilderTest.kt

malinajirka commented on WordPress/src/test/java/org/wordpress/android/ui/jetpack/scan/ScanViewModelTest.kt:
```diff
@@ -517,22 +518,23 @@ class ScanViewModelTest : BaseUnitTest() {
-
-            assertThat(indeterminateProgressBars.isNotEmpty()).isTrue
-        }
+    // TODO ashiagr fix test
```
Is this intentionally commented out or were you planning on getting back to it in this PR?

ashiagr replied:
Oh yes, I wanted to fix it here. Thanks for pointing it.

I've removed vm test cases related to updating button and progress bar on threats fixing in 7e1742e as threats fixing content building is now done in the builder class.

malinajirka commented on WordPress/src/test/java/org/wordpress/android/ui/jetpack/scan/builders/ScanStateListItemsBuilderTest.kt:
```diff
@@ -60,8 +68,8 @@ class ScanStateListItemsBuilderTest : BaseUnitTest() {
+            threatDetailsListItemsBuilder,
+            scanStore
         )
 //        whenever(htmlMessageUtils.getHtmlMessageFromStringFormatResId(anyInt(), any())).thenReturn(SpannedString(""))
```
❓ Could we remove this commented out code as well? We can always find it in the git history if needed. Wdyt?

ashiagr replied:
Yes, removed in d61a3e8.

malinajirka reviewed (approved):
Looks great @ashiagr, thanks!

ashiagr committed:
Remove tests for scan state content items for fixing threats state

 There's now a separate threats fixing state in the scan state builder class and tests for threats fixing state content items are added to builder test class.

ashiagr committed:
Remove commented test code

ashiagr commented:
Thanks for the review @malinajirka, ready for another look.

zwarm committed:
Merge pull request #13937 from wordpress-mobile/issue/13326-here-to-help

Issue/13326 here to help

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/jetpack/scan/builders/ScanStateListItemsBuilder.kt:
```diff
@@ -242,16 +246,28 @@ class ScanStateListItemsBuilder @Inject constructor(
+            )
+
+        val clickableText = resourceProvider.getString(R.string.scan_here_to_help)
+        val startIndex = descriptionText.indexOf(clickableText)
```
⚠️ I'm worried that this might not work as expected. No-one can guarantee that translations to other languages will have the exact same translation for `scan_here_to_help` as what is used in `scan_idle_threats_found_description` -> so I think this might return -1. Or am I missing something?


What if we used a string argument instead of relying on the translation. Wdyt?

```
<string name="scan_idle_threats_found_description">The scan found %1$s potential threats on %2$s. Please review them below and take action. We\'re %3$s if you need us.</string>
<string name="scan_here_to_help">here to help</string>
```

Wdyt?

ashiagr replied:
Oops, yes you're right....will fix it shortly.

ashiagr replied:
Fixed in 6dbdce9.

malinajirka reviewed (changes requested):
Thanks for the improvements @ashiagr! I found one potential issue, let me know what you think ;).

ashiagr committed:
Fix: Pass clickable text as string argument into description text

ashiagr committed:
Fix lint

malinajirka reviewed (approved):
LGTM, thanks!
